ts/packages/ts/src/commands/UpsertCommand.ts

```diff
@@ -5,59 +5,94 @@ import { ConfigError } from '../utils';
 import { type ConfigStore } from '../ConfigStore';
 import { type ConfigSet } from '../ConfigSet';
 import { ConfigSchema } from '../ConfigSchema';
+import { type EvalCommandReturn } from './EvalCommand';
 
 export type UpsertCommandParameters = {
   store: ConfigStore;
   set: ConfigSet;
   schema: ConfigSchema;
-  configs: { [key: string]: string };
+  configs?: { [key: string]: string };
+  pipe?: EvalCommandReturn;
 };
 
 export class UpsertCommand extends Command<void> {
   constructor(public parameters: UpsertCommandParameters) {
     super(parameters);
   }
 
-  async run() {
-    const { store, set, schema, configs } = this.parameters;
+  private validateConfigValue({
+    key,
+    value,
+    skipDeclarationAndTemplateValidation,
+  }: {
+    key: string;
+    value: string;
+    skipDeclarationAndTemplateValidation: boolean;
+  }) {
+    const { store, set, schema } = this.parameters;
 
-    await store.init();
+    const errorScope: [string, string][] = [
+      ['UpsertCommand', `store:${store.type};set:${set.path};schema:${schema.name}`],
+      ['parameters.configs', `key:${key};value:${value}`],
+    ];
 
-    const upsertConfigs = _(configs)
-      .entries()
-      .map<Config>(([key, value]) => {
-        const errorScope: [string, string][] = [
-          ['UpsertCommand', `store:${store.type};set:${set.path};schema:${schema.name}`],
-          ['parameters.configs', `key:${key};value:${value}`],
-        ];
+    const cfgu = schema.contents[key];
 
-        const cfgu = schema.contents[key];
+    if (!skipDeclarationAndTemplateValidation) {
+      if (!cfgu) {
+        throw new ConfigError(
+          'invalid config key',
+          `key "${key}" must be declared on schema ${schema.name}`,
+          errorScope,
+        );
+      }
+      if (value !== undefined && cfgu?.template) {
+        throw new ConfigError('invalid config value', `keys declared with template mustn't have a value`, errorScope);
+      }
+    }
 
-        if (!cfgu) {
-          throw new ConfigError(
-            'invalid config key',
-            `key "${key}" must be declared on schema ${schema.name}`,
-            errorScope,
-          );
-        }
+    if (value !== undefined && cfgu) {
+      try {
+        ConfigSchema.CFGU.VALIDATORS.valueOptions(cfgu, value);
+        ConfigSchema.CFGU.VALIDATORS.valueType(cfgu, value);
+      } catch (error) {
+        throw error?.appendScope?.(errorScope) ?? error;
+      }
+    }
+  }
+
+  async run() {
+    const { store, set, schema, configs, pipe } = this.parameters;
 
-        if (value) {
-          if (cfgu.template) {
-            throw new ConfigError(
-              'invalid config value',
-              `keys declared with template mustn't have a value`,
-              errorScope,
-            );
-          }
+    if (_.isEmpty(configs) && _.isEmpty(pipe)) {
```

Comment:
why not sppliting it via wach one of the loops?

Reply:
I don't understand the question. This condition is to skip the async init method in case there is nothing to upsert.